pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -933,7 +943,6 @@ protected void doTakeSnapshot() {
           continue;
         }
         immutableSegment.persistValidDocIdsSnapshot();
-        _updatedSegmentsSinceLastSnapshot.remove(segment);
```

Comment:
nit: we can still keep this `remove` statement. If snapshotting errors out in between then we don't need to process the segments for snapshot in next cycle again if they were processed this cycle.

Reply:
Right, if it keeps erroring out and the set is not cleared, then many segments would be taken snapshots repeatedly. I’d assume such error case would be rare and if it happened things are not getting worse than before when all segments were take snapshots all the time. I’ll probably skip this optimization for now and revisit when we see need. But good point anyway.